### ISSUE TITLE
Training the G2P on latest version is much slower than 2.0.0a4

### ISSUE THREAD

**Issue by george-roussos** [MontrealCorpusTools/Montreal-Forced-Aligner#289]
I have been trying to reproduce a train run similar to one that I performed using `2.0.0a4` (since models trained on that version are not compatible with latest). The exact same configuration on latest version is giving me very much slower train times. I tried to confirm that it wasn't anything on my end, so I created a `conda` environment on `2.0.0a4` and, the same configuration on that older version can train an FST within 5 hours -- with latest it is giving an estimate of 48 hours. Is this something known/expected and if so why? I took a look at `g2p/trainer.py` and can see quite a number of changes (e.g. now `delta` and `lr` is mentioned in the train script).

For context, the configuration I want to run looks like `mfa train_g2p lexicon_txt.txt dictionary/7gram.zip --order 7 --v --j 8`

**Comment by mmcauliffe** [comment]
No, it shouldn't be that much slower, there was an update for Pynini versions to the latest, but I'll dive in and see what's causing the slow down.

**Comment by mmcauliffe** [comment]
One thing is that the argument for number of jobs should be specified as `-j 8` or `--num_jobs 8`, otherwise it'll stay at the default 3, but I'm assuming that you're running the same command in both cases, so I'll still dive further into it.

**Comment by george-roussos** [comment]
Oh yeah, options/flags, my bad, so it is not taken into account. Although yeah, I ran the same configuration using the older version and it output much less time left when calculating the alignments, even without the option (well, the wrong one I gave). I will give it a go, though.

Thanks for looking into it, appreciate it. Can I take the chance to ask if there is a way to get verbose during the training?

**Comment by george-roussos** [comment]
Oh yeah, my bad, so it is not taken into account. Although yeah, I ran the same configuration using the older version and it output much less time left when calculating the alignments, even without the option (well, the wrong one I gave). I will give it a go, though.

I now see I did try `-j` too, but that was locally, on my laptop.

Thanks for looking into it, appreciate it. Can I take the chance to ask if there is a way to get verbose during the training?

**Comment by mmcauliffe** [comment]
Yeah, I can add some better output, the G2P uses a different logging set up than the rest that I haven't standardized.  

One thing to note about the progress bar and the estimated time is that it's not very accurate at the beginning particularly. The progress bar doesn't take into account information about multiple processes, so the initial couple are not going to be great estimation, especially with lots of processes.

It does look like the `baumwelchtrain` from the most recent version is significantly slower than the version that 2.0.0a4 was using, so I'll see if that's intended.  The recent updates added a stepwise training with a learning rate, so my guess is that it takes longer but results in better G2P models at the end.

**Comment by george-roussos** [comment]
Oh thanks! Good to know (I also thought it might be on `baumwelchtrain`). I saw the changes in the trainer script involving `delta` and `lr` and I assumed that the changes might in fact mean improved performance, but jury's still out so we'll have to see. I guess if longer train times cannot be avoided, live output during training may be a good idea. Thanks again for the great library. 🙂👍🏻

**Comment by george-roussos** [comment]
Hi again, now I have tried another run with `-j 6` and `-j 8` (I have 8 cores) but I did not notice any improvements -- in fact, I left it running all night, but it was even slower (did not even move to `25%`). My guess is that more jobs are more of a bottleneck. I did try to run the command with `OMP_NUM_THREADS=1` (it relieves problems I have with Torch's `dataloader` sometimes), but that didn't do the trick either. I guess we are out of luck if `baumwelch `now only works like this.

**Comment by mmcauliffe** [comment]
Should be fixed in the latest version.  I set the batch_size to 200, which works pretty well for my case, but you can play around with it via `--batch_size`.  Also I put in arguments for `max_iterations` and `learning_rate` if you want to play around with those too, but the batch size should fix the performance issues.

**Comment by george-roussos** [comment]
Oh thanks, that's very kind of you. I will play around and see what I can get out of it.

**Comment by mmcauliffe** [comment]
Yeah, so it does a validation pass training on 90% and testing on 10% and then trains again on the whole dictionary.  I agree that's not super intuitive, so I might change that, since the default behavior without the flag is to do training on the whole dictionary, so with the flag it should just do the 90% training and output that model.

**Comment by george-roussos** [comment]
Oh cool thanks, cool to know. Thanks for the library, it's neat! 👍🏻
